parlai/agents/starspace_torch/starspace_torch.py

```diff
@@ -0,0 +1,367 @@
+            ye.view(-1, xe.size(-1)),
+            y.view(-1)
+        )
+        loss.backward()
+        self.optimizer.step()
+        scores = nn.CosineSimilarity(dim=-1).forward(xe, ye)
+
+        # Update metrics
+        self.metrics['loss'] += loss.item()
+        self.metrics['examples'] += batchsize
+        _, ranks = scores.sort(1, descending=True)
+        for b in range(batchsize):
+            rank = (ranks[b] == label_inds[b]).nonzero().item()
+            self.metrics['rank'] += 1 + rank
+
+        return Output()
+
+    def eval_step(self, batch):
+        """Evaluate a single batch of examples."""
+        if batch.text_vec is None:
+            return
+        batchsize = batch.text_vec.size(0)
+        self.model.eval()
+
+        cands, cand_vecs, label_inds = self._build_candidates(
+            batch, source=self.opt['eval_candidates'], mode='eval')
+
+        xe, ye = self.model(
+            batch.text_vec,
+            ys=None,
+            cands=cand_vecs
+        )
+        xe = torch.cat([xe for _ in range(ye.size(1))], dim=1)
+
+        scores = nn.CosineSimilarity(dim=-1).forward(xe, ye)
+        _, ranks = scores.sort(1, descending=True)
+
+        self.metrics['examples'] += batchsize
+        for b in range(batchsize):
+            rank = (ranks[b] == label_inds[b]).nonzero().item()
+            self.metrics['rank'] += 1 + rank
+
+        cand_preds = []
+        for i, ordering in enumerate(ranks):
+            if cand_vecs.dim() == 2:
+                cand_list = cands
+            elif cand_vecs.dim() == 3:
+                cand_list = cands[i]
+            cand_preds.append([cand_list[rank] for rank in ordering])
+        preds = [cand_preds[i][0] for i in range(batchsize)]
+        return Output(preds, cand_preds)
+
+    def observe(self, observation):
+        labels = observation.get('labels')
+        if not labels:
+            labels = observation.get('eval_labels')
+        if labels:
+            for label in labels:
+                self.add_to_ys_cache(label)
+        return super(StarspaceTorchAgent, self).observe(observation)
+
+    def vectorize(self, obs, add_start=False, add_end=False, truncate=None,
+                  split_lines=False):
+        """Overrride method from Torch Agent so that by default we do not add
+        start and end indices."""
+        return super(StarspaceTorchAgent, self).vectorize(
+            obs,
+            add_start=add_start,
+            add_end=add_end,
+            truncate=truncate,
+            split_lines=split_lines,
+        )
+
+    def _get_custom_candidates(self, obs):
+        """Sets custom candidates for the observation."""
+        if not obs.get('eval_labels'):
+            # only use custom candidates during training
+            cands = []
+            cache_sz = len(self.ys_cache) - 1
+            if cache_sz < 1:
+                return cands
+            k = self.opt['neg_samples']
+            for i in range(1, k * 3):
```

Comment:
Why 3?

Reply:
that was what was in original starspace

Reply:
i'm guessing it was just a "reasonable" number of examples to search through to get k negatives that were not equal to the label